src/idatetime.c

```diff
@@ -0,0 +1,186 @@
+#include "data.table.h"
+
+#define YEARS400 146097
+#define YEARS100 36524
+#define YEARS4 1461
+#define YEARS1 365
+
+typedef enum { YDAY, WDAY, MDAY, WEEK, MONTH, QUARTER, YEAR, YEARMON, YEARQTR} datetype;
+
+int convertDate_int(int x, datetype type);
+double convertDate_double(int x, datetype type);
+
+SEXP convertDate(SEXP x, SEXP type)
+{
+    if (!isInteger(x)) error(_("x must be an integer vector"));
+    const int *ix = INTEGER(x);
+    const int n = length(x);
+    if (!isString(type) || length(type) != 1)
+        error(_("Internal error: invalid type for convertDate(), should have been caught before. please report to data.table issue tracker")); // # nocov
+    datetype ctype;
+    bool ansint = true;
+    if (!strcmp(CHAR(STRING_ELT(type, 0)), "yday")) ctype = YDAY;
+    else if (!strcmp(CHAR(STRING_ELT(type, 0)), "wday")) ctype = WDAY;
+    else if (!strcmp(CHAR(STRING_ELT(type, 0)), "mday")) ctype = MDAY;
+    else if (!strcmp(CHAR(STRING_ELT(type, 0)), "week")) ctype = WEEK;
+    else if (!strcmp(CHAR(STRING_ELT(type, 0)), "month")) ctype = MONTH;
+    else if (!strcmp(CHAR(STRING_ELT(type, 0)), "quarter")) ctype = QUARTER;
+    else if (!strcmp(CHAR(STRING_ELT(type, 0)), "year")) ctype = YEAR;
+    else if (!strcmp(CHAR(STRING_ELT(type, 0)), "yearmon")) { ctype = YEARMON; ansint = false; }
+    else if (!strcmp(CHAR(STRING_ELT(type, 0)), "yearqtr")) { ctype = YEARQTR; ansint = false; }
+    else error(_("Internal error: invalid type for convertDate, should have been caught before. please report to data.table issue tracker")); // # nocov
+
+    SEXP ans;
+    if (ansint) {
+        ans = PROTECT(allocVector(INTSXP, n));
+        int *ansp = INTEGER(ans);
+        for (int i=0; i < n; ++i) {
+            ansp[i] = convertDate_int(ix[i], ctype);
+        }
+    } else {
+        ans = PROTECT(allocVector(REALSXP, n));
+        double *ansp = REAL(ans);
+        for (int i=0; i < n; ++i) {
+            ansp[i] = convertDate_double(ix[i], ctype);
+        }
+    }
+    UNPROTECT(1);
+    return ans;
+}
+
+int convertDate_int(int x, datetype type)
```

Comment:
it looks like there's a lot of duplication between `convertDate_int` and `convertDate_double` -- could they be combined? C++ templates would help a lot here... `void convertSingleDate(int x, double *double_out, int *int_out, is_double, datetype type)` might work?

Otherwise, use more helper functions to unify shared logic across the two functions...

Reply:
Indeed. I also thought about either passing result pointers or using a struct where the results are stored, although the pointers seem favorable in terms of speed. Will rewrite it the next days.